File: leetcode/leetcode/729_My_Calendar_I.cpp
//
//  729_My_Calendar_I.cpp
//  leetcode
//
//  Created by Hang Zhang on 4/21/20.
//  Copyright © 2020 Hilings Studio. All rights reserved.
//

#include <iostream>
#include <map>

using namespace std;

class MyCalendar {
    map<int, int> om;

  public:
    MyCalendar() {
    }

    bool book(int start, int end) {
        auto it = om.lower_bound(start);
        if ((it == om.begin() || prev(it)->second <= start) && (it == om.end() || end <= it->first)) {
            om[start] = end;
            printf("[%d, %d] true\n", start, end);
            return true;
        }
        printf("[%d, %d] false\n", start, end);
        return false;
    }

    void print() {
        for (auto p : om) {
            cout << "[" << p.first << ", " << p.second << "]\n";
        }
    }
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 729. My Calendar I, C++ ...\n\n";

    MyCalendar sol = MyCalendar();
    // sol.book(10, 20); // returns true
    // sol.book(15, 25); // returns false
    // sol.book(20, 30); // returns true

    sol.book(47, 50); // true
    sol.book(33, 41); // true
    sol.book(39, 45); // false
    sol.book(33, 42); // false
    sol.book(25, 32); // true
    sol.book(26, 35); // false
    sol.book(19, 25); // true
    sol.book(3, 8);   // true
    sol.book(8, 13);  // true
    sol.book(18, 27); // false

    // sol.book(37, 50); // true
    // sol.book(33, 50); // false
    // sol.book(4, 17);  // true
    // sol.book(35, 48); // false
    // sol.book(8, 25);  // false

    sol.print();

    return 0;
}

File: go/729_My_Calendar_I.go
package main

type MyCalendar struct {
	Books [][]int
}

func Constructor() MyCalendar {
	mc := MyCalendar{
		Books: [][]int{},
	}
	return mc
}

func (this *MyCalendar) Book(start int, end int) bool {
	for _, book := range this.Books {
		if end <= book[0] || start >= book[1] {
			continue
		}
		return false
	}
	this.Books = append(this.Books, []int{start, end})
	return true
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Book(start,end);
 */

// func main() {
// 	fmt.Println("LeetCode 729. My Calendar I ...")

// 	mc := Constructor()
// 	r := mc.Book(10, 20)
// 	fmt.Println(r)

// 	r = mc.Book(15, 25)
// 	fmt.Println(r)

// 	r = mc.Book(20, 30)
// 	fmt.Println(r)
// }
